aten/src/ATen/core/ivalue.h

```diff
@@ -751,35 +751,33 @@ DEFINE_TO(c10::Device, toDevice)
 DEFINE_TO(at::ScalarType, toScalarType)
 DEFINE_TO(at::Layout, toLayout)
 
-template<typename Elem>
+template <typename Elem>
 inline std::vector<Elem> to_list(std::vector<IValue>& list) {
-  std::vector<Elem> converted_list;
-  std::transform(
-      list.begin(), list.end(), std::back_inserter(converted_list), [
-      ](IValue ivalue) -> Elem {
-        return ivalue.to<Elem>();
-      });
-  return std::move(converted_list);
+  return fmap(list, [](IValue ivalue) { return ivalue.to<Elem>(); });
 }
 
-template<typename Elem>
+template <typename Elem>
 inline std::vector<Elem> to_list(const std::vector<IValue>& list) {
-  std::vector<Elem> converted_list;
-  std::transform(
-      list.begin(), list.end(), std::back_inserter(converted_list), [
-      ](IValue ivalue) -> Elem {
-        return ivalue.to<Elem>();
-      });
-  return std::move(converted_list);
+  return fmap(list, [](IValue ivalue) { return ivalue.to<Elem>(); });
 }
 
-template<typename T>
+template <typename Elem>
+struct is_vector : std::false_type {};
+
+template <typename Elem, typename Allocator>
+struct is_vector<std::vector<Elem, Allocator>> : std::true_type {};
+
+template <typename T>
 inline T IValue::to() && {
+  static_assert(
```

Comment:
tbh I was thinking more like, check isGenericListRef() and isGenericDictRef() or throw.

btw: is there a reason why you're not adding GenericDict here? Seems like it would be good to add.

Reply:
static assert is more friendly since it should give an error at compile time rather than runtime if the types aren't right, and the error message info available isn't much better at runtime

I'll rebase and add dictionaries after #17175 goes in